### ISSUE TITLE
Warning: a promise was created in a handler but was not returned from it

### ISSUE THREAD

**Issue by pocesar** [jspm/jspm-cli#1956]
```
Warning: a promise was created in a handler but was not returned from it
    at depsCallback (C:\nodejs\node_modules\jspm\lib\install.js:504:59)
    at doDepsCallback (C:\nodejs\node_modules\jspm\lib\package.js:410:5)
    at C:\nodejs\node_modules\jspm\lib\package.js:419:7
    at Array.forEach (native)
    at preloadResolve (C:\nodejs\node_modules\jspm\lib\package.js:418:22)
    at C:\nodejs\node_modules\jspm\lib\package.js:579:13
    at C:\nodejs\node_modules\jspm\node_modules\graceful-fs\graceful-fs.js:78:16
From previous event:
    at C:\nodejs\node_modules\jspm\lib\package.js:575:12
From previous event:
    at C:\nodejs\node_modules\jspm\lib\package.js:570:8
From previous event:
    at C:\nodejs\node_modules\jspm\lib\package.js:518:6
From previous event:
    at Object.exports.download (C:\nodejs\node_modules\jspm\lib\package.js:452:4)
    at C:\nodejs\node_modules\jspm\lib\install.js:408:18
    at tryOnImmediate (timers.js:534:15)
    at processImmediate [as _immediateCallback] (timers.js:514:5)
From previous event:
    at C:\nodejs\node_modules\jspm\lib\install.js:404:6
From previous event:
    at install (C:\nodejs\node_modules\jspm\lib\install.js:387:4)
    at C:\nodejs\node_modules\jspm\lib\install.js:142:14
    at Array.map (native)
    at C:\nodejs\node_modules\jspm\lib\install.js:131:45
```

this error appears constantly when doing an `jspm install`, for each dependency. using jspm 0.17.0-beta.22, node 5.12

**Comment by guybedford** [comment]
It sounds like you've turned on warnings in Bluebird?

The warning sounds like it's there to prevent memory leaks, but we actually use this functionality for a purpose.

**Comment by pocesar** [comment]
indeed, my NODE_ENV is set to development, and it works the same as BLUEBIRD_DEBUG. 

bluebird recomend to end the last `then` of a promise chain to `return null` if it's intended for side-effects (acting like callbacks)

**Comment by guybedford** [comment]
@pocesar thanks do you mean like https://github.com/jspm/jspm-cli/commit/7e219c5d31a5f0c2a833293cedc6b24a84c21df0 ?

**Comment by pocesar** [comment]
@guybedford exactly as that as stated in [bluebird docs](http://bluebirdjs.com/docs/warning-explanations.html#warning-a-promise-was-created-in-a-handler-but-was-not-returned-from-it)

**Comment by m-a-r-c-e-l-i-n-o** [comment]
@pocesar @guybedford Did https://github.com/jspm/jspm-cli/commit/7e219c5d31a5f0c2a833293cedc6b24a84c21df0 resolve the issue for you guys? I am dealing with it right now and the patch didn't do anything for me. Also tried other variations and nothing. Perhaps it's worth investigating a bit more? Issue is present in development mode on jspm `v0.17.0-beta.22` with node `v5.12.0`, doing `JSPM.install(true, { lock: true })`.

**Comment by guybedford** [comment]
The original fix I added here was incorrect - I added a possible correction in https://github.com/jspm/jspm-cli/commit/56bf63f245a543e9d8c35ba429a7a41de5afa5ec. I still haven't been able to replicate this though even explicitly setting warnings to true with Bluebird 3.4.0 but let me know if this works ok here.

**Comment by glen-84** [comment]
@guybedford Did you try with `NODE_ENV` set to `development`?

**Comment by guybedford** [comment]
@glen-84 yes we do reject with strings in various places.

**Comment by pocesar** [comment]
@guybedford bluebird docs also says it's good practice to reject errors (since the library won't be needing to synthetically recreate the stack). not sure the implications for jspm though, like if it relies on the string given to act accordingly. rejecting with an error is also good when you're filtering errors in the `catch()` callback, like so:

```js
somePromise
.catch('JspmRequestError' /* error.name */, function(){ /* will catch only this type of error*/ })
.catch('ReferenceError' /* error.name */, function(){ /* will catch only this type of error*/ })
.catch(function(){ /* catch-all but thoese 2 above */ })
```

**Comment by m-a-r-c-e-l-i-n-o** [comment]
Just wanted to give an update, I did try both patches, with no conclusive results. The first did not make a difference, so I moved forward with `process.env['NODE_ENV'] = 'testing'`, as oppose to `'development'`.  Two weeks worth of commits later (with no jspm upgrades), I'm now switching over to `process.env['NODE_ENV'] = 'development'`. I still see a bluebird warning `Warning: .then() only accepts functions but was passed: [object Undefined], [object Undefined]`, just not directly related jspm. It's coming from `/node_modules/karma/lib/middleware/common.js:14:39`. 

I can't completely rule out  jspm because I do use a custom fork of [karma-jspm](https://github.com/Workiva/karma-jspm) in my Karma setup, so jspm is in the midst. The warning shows up with or without the patches @guybedford proposed. I tried them individually as well.

I don't have the time to try to isolate this atm, it's a very robust project, otherwise I would, but the takeaway here is that perhaps this isn't a jspm issue. I mean, @guybedford is having a hard time replicating it, and my direct jspm warnings seem to be gone without jspm upgrades in my setup.

Btw, it appears that you can force BlueBird to shutup by setting `process.env['BLUEBIRD_WARNINGS'] = 0`, it's a viable solution for my setup, so it'll do for now.

**Comment by pocesar** [comment]
I'm setting the console var with BLUEBIRD_DEBUG=0, it makes jspm update/install "silent", since I'm still seeing this issue in beta-24

**Comment by guybedford** [comment]
I will try again to replicate, but it didn't replicate last time I tried.
If you can share any steps from a new project to get the issue that would
certainly help.

**Comment by guybedford** [comment]
I finally managed to replicate this - it seems this is actually a windows-only bug.

**Comment by pocesar** [comment]
@guybedford how so? something to do with ENV vars?

**Comment by guybedford** [comment]
Fixed in https://github.com/jspm/jspm-cli/commit/65df6a0875cb6762a5fe0d68681dd525f7003e37.

**Comment by guybedford** [comment]
(I'm still not quite sure why this only came up in windows for me though)

**Comment by glen-84** [comment]
@guybedford Maybe you have some configuration or environment variables on your non-Windows machine that are disabling the warnings?

**Comment by guybedford** [comment]
Fixed in the latest 0.17 release.
